iree/compiler/Dialect/VM/Transforms/test/sink_defining_ops.mlir

```diff
@@ -0,0 +1,70 @@
+// RUN: iree-opt -split-input-file -iree-vm-sink-defining-ops %s | IreeFileCheck %s
+
+vm.module @module {
+  // CHECK-LABEL: @single_uses
+  vm.func @single_uses(%arg0 : i32) -> i32 {
+    %c4 = vm.const.i32 4 : i32
+    %c5 = vm.const.i32 5 : i32
+    // CHECK: vm.cond_br
+    vm.cond_br %arg0, ^bb1, ^bb2
+  ^bb1:
+    //      CHECK: %c4 = vm.const.i32 4 : i32
+    // CHECK-NEXT: vm.return %c4 : i32
+    vm.return %c4 : i32
+  ^bb2:
+    //      CHECK: %c5 = vm.const.i32 5 : i32
+    // CHECK-NEXT: vm.return %c5 : i32
+    vm.return %c5 : i32
+  }
+}
+
+// -----
+
+vm.module @module {
+  // CHECK-LABEL: @multiple_uses
+  vm.func @multiple_uses(%arg0 : i32) -> (i32, i32) {
+    %c4 = vm.const.i32 4 : i32
+    %c5 = vm.const.i32 5 : i32
+    // CHECK: %c5 = vm.const.i32 5 : i32
+    // CHECK: vm.cond_br
+    vm.cond_br %arg0, ^bb1, ^bb2
+  ^bb1:
+    //      CHECK: %c4 = vm.const.i32 4 : i32
+    // CHECK-NEXT: vm.return %c4, %c5
+    vm.return %c4, %c5 : i32, i32
+  ^bb2:
+    // CHECK: vm.return %c5 : i32
+    vm.return %c5 : i32
+  }
+}
+
+// -----
+
+vm.module @module {
+  // CHECK-LABEL: @common_dominator
+  vm.func @common_dominator(%arg0 : i32, %arg1 : i32) -> (i32, i32) {
+    %c4 = vm.const.i32 4 : i32
+    %c5 = vm.const.i32 5 : i32
+    %c6 = vm.const.i32 6 : i32
+    //      CHECK: %c5 = vm.const.i32 5 : i32
+    // CHECK-NEXT: vm.cond_br %arg0
+    vm.cond_br %arg0, ^bb1, ^bb_end
+  ^bb1:
+    //      CHECK: "test.do_something_else"
+    "test.do_something_else"() : () -> ()
+    // CHECK-NEXT: %c4 = vm.const.i32 4 : i32
```

Comment:
Why is `%c4` here in `^bb1`, when it is also used in `^bb2`? Is it defined in both places?

Reply:
^bb1 dominates ^bb2 (all code paths leading to ^bb2 go through ^bb1), so the value can be used there. this pass tries to move the values down to the first use.

Reply:
Ah, I missed that property of `^bb2`. Cool!